sirepo/template/srw.py

```diff
@@ -1898,17 +1898,39 @@ def _remap_3d(info, allrange, z_label, z_units, width_pixels, scale='linear'):
         except Exception:
             pkdlog('Cannot resize the image - scipy.ndimage.zoom() cannot be imported.')
             pass
-    z_label = z_label
-    if z_units:
-        z_label += ' [' + z_units + ']'
+    # rotate 3D image    
+    if rotate_angle:
+        rotate_reshape = (rotate_reshape == "1")
+        try:
+            from scipy import ndimage
+            pkdlog('Size before: {}  Dimensions: {}', ar2d.size, ar2d.shape)
+            shape_before = list(ar2d.shape)
+            ar2d = ndimage.rotate(ar2d, rotate_angle, reshape = rotate_reshape, mode='constant', order = 3)
+            pkdlog('Size after rotate: {}  Dimensions: {}', ar2d.size, ar2d.shape)
+            shape_rotate = list(ar2d.shape)
+
+            pkdlog('x_range and y_range before rotate is [{},{}] and [{},{}]', x_range[0], x_range[1], y_range[0], y_range[1])
+            x_range[0] = shape_rotate[0]/shape_before[0]*x_range[0]
+            x_range[1] = shape_rotate[0]/shape_before[0]*x_range[1]
+            y_range[0] = shape_rotate[1]/shape_before[1]*y_range[0]
+            y_range[1] = shape_rotate[1]/shape_before[1]*y_range[1]
+            pkdlog('x_range and y_range after rotate is [{},{}] and [{},{}]', x_range[0], x_range[1], y_range[0], y_range[1])
+
+            x_range[2] = ar2d.shape[1]
+            y_range[2] = ar2d.shape[0]
+            if info['title'] != 'Power Density': info['subtitle'] = info['subtitle'] + ' Image Rotate {}^0'.format(rotate_angle)
+        except Exception:
+            pkdlog('Cannot rotate the image - scipy.ndimage.rotate() cannot be imported.')
```

Comment:
This is the correct place to use pkdlog() - it indicates that something is wrong, but the software is continuing anyway. On a production server, this message would appear in the logs and indicate there is a problem.

Reply:
Get it! Thanks a lot!